File: packages/java/src/main/java/com/github/ponstream24/orbitid/OrbitError.java
package com.github.ponstream24.orbitid;

/** Exception raised when an Orbit ID value or generator operation is invalid. */
public final class OrbitError extends RuntimeException {
    public static final String INVALID_TYPE = "INVALID_TYPE";
    public static final String INVALID_NODE = "INVALID_NODE";
    public static final String INVALID_SEQUENCE = "INVALID_SEQUENCE";
    public static final String INVALID_TIMESTAMP = "INVALID_TIMESTAMP";
    public static final String INVALID_DECIMAL = "INVALID_DECIMAL";
    public static final String CLOCK_ROLLBACK = "CLOCK_ROLLBACK";
    public static final String SEQUENCE_EXHAUSTED = "SEQUENCE_EXHAUSTED";
    public static final String NODE_OWNERSHIP_LOST = "NODE_OWNERSHIP_LOST";

    /** Stable machine-readable code matching the TypeScript implementation. */
    public final String code;

    public OrbitError(String code, String message) {
        super(message);
        this.code = code;
    }

    public String getCode() {
        return code;
    }
}

File: packages/php/src/OrbitError.php
<?php

declare(strict_types=1);

namespace OrbitId;

final class OrbitError extends \RuntimeException
{
    public const INVALID_TYPE = 'INVALID_TYPE';
    public const INVALID_NODE = 'INVALID_NODE';
    public const INVALID_SEQUENCE = 'INVALID_SEQUENCE';
    public const INVALID_TIMESTAMP = 'INVALID_TIMESTAMP';
    public const INVALID_DECIMAL = 'INVALID_DECIMAL';
    public const CLOCK_ROLLBACK = 'CLOCK_ROLLBACK';
    public const SEQUENCE_EXHAUSTED = 'SEQUENCE_EXHAUSTED';
    public const NODE_OWNERSHIP_LOST = 'NODE_OWNERSHIP_LOST';

    public readonly string $orbitCode;

    public function __construct(string $orbitCode, string $message)
    {
        $this->orbitCode = $orbitCode;
        parent::__construct($orbitCode . ': ' . $message);
    }
}
